New upstream release tool

mattclay opened the pull request:
##### SUMMARY

New upstream release tool.

##### ISSUE TYPE

Feature Pull Request

##### COMPONENT NAME

packaging/release.py

##### ADDITIONAL INFORMATION

I looked at quite a few Python CLI frameworks, but finally settled on implementing something on top of `argparse`. Not having an extra requirement was nice, but the main reason was that none of the frameworks offered quite what I was looking for.

I wanted a simple, low boilerplate framework that allowed easily chaining commands together, like the `notify` method in `nox`. I also wanted support for global activation of argcomplete. Often this wasn’t supported, and some frameworks had no argcomplete support or even lacked support for tab completion entirely.

I tried to keep the tool flexible, while minimizing the risk of making mistakes during a release. After discussing the workflow briefly with @nitzmahone I made sure to include support for working with a detached HEAD.

The requirements to run the tool are:

* jinja2
* packaging

All other requirements are installed using virtual environments managed by the tool.

Aside from the `twine upload`, all other steps go through human review in the user’s browser.

Publishing releases requires the following access:

* Permission to merge PRs and create GitHub releases for ansible/ansible.
* Permission to upload to PyPI.
* Permission to restart AZP jobs (in case of transient CI failures).
* Email (it is recommended that your `mailto:` handler works).

I’ve done some end-to-end testing of the release process using this tool. I’ve provided links to some of the results:

**2.14.4**

https://github.com/ansible/test-ansible-release/pull/1
https://github.com/ansible/test-ansible-release/pull/2
https://test.pypi.org/project/ansible-core/2.14.4/
https://github.com/ansible/test-ansible-release/releases/tag/v2.14.4

**2.14.5rc1**

https://github.com/ansible/test-ansible-release/pull/3
https://github.com/ansible/test-ansible-release/pull/4
https://test.pypi.org/project/ansible-core/2.14.5rc1/
https://github.com/ansible/test-ansible-release/releases/tag/v2.14.5rc1

mattclay committed:
New upstream release tool

mattclay commented:
Here's the current status of support for the stable branches:

* stable-2.14 - Supported
* stable-2.13 - Blocked on https://github.com/ansible/ansible/pull/80130
* stable-2.12 - Blocked on https://github.com/ansible/ansible/pull/80131
* stable-2.9 - Not supported

webknjaz commented on packaging/release.py:
```diff
@@ -0,0 +1,1410 @@
+
+from packaging.version import Version, InvalidVersion
+
+# region CLI Framework
```
This thing is quite huge. Any chance to split it into pieces and make it easier to digest?

mattclay replied:
I kept it in a single file for convenience while testing across the various branches and in various scenarios, since it's easier to copy changes around when they're all in one file. I agree it is a bit large, but I don't see it growing much larger, so I'd rather leave it as-is for now.

webknjaz commented on packaging/release.py:
```diff
@@ -0,0 +1,1410 @@
+
+    ansible_requirements = ANSIBLE_REQUIREMENTS_FILE.read_text()
+
+    release_requirements = """
```
Why keep this as a string literal? A tuple would read better, it can be joined later on.

mattclay replied:
It was convenient, since all the other requirements are also strings. It didn't seem any more readable when split, and this way avoids having to add a join.

webknjaz commented on packaging/release.py:
```diff
@@ -0,0 +1,1410 @@
+    wheel_file = get_wheel_path(version)
+    env = ensure_venv()
+
+    if prompt:
```
I think this would be unusable from GHA since it'd require using the action with OIDC support instead.

mattclay replied:
The `--no-prompt` option for this command will allow it to be used non-interactively, if required.

webknjaz replied:
@mattclay that's not enough. It should also *NOT* run `twine upload` at all.

mattclay replied:
If the tool is being used in a workflow where a command, such as `publish`, isn't desired, then the command should not be used.

webknjaz replied:
Okay, that'll do, I guess. But then, my initial suggestion stands — if there's no need for two modes, why implement both?

mattclay replied:
The option was added in case someone wanted to bypass the prompt while using twine.

webknjaz commented on packaging/release.py:
```diff
@@ -0,0 +1,1410 @@
+            except FileNotFoundError:
+                raise ApplicationError(f"Missing sdist: {sdist_file.relative_to(CHECKOUT_DIR)}") from None
+
+            sdist.extractall(temp_dir)
```
What does this test, and why not use the `tarfile`'s APIs for checking the file paths w/o actually unpacking to dist?

mattclay replied:
It verifies the file can be extracted before testing with `pip install`. The current checks on the extracted files could have been done with the `tarfile` API, but using `glob` and `rglob` were a simple and convenient way to perform the checks.

webknjaz commented on packaging/release.py:
```diff
@@ -0,0 +1,1410 @@
+    with tempfile.TemporaryDirectory() as temp_dir_name:
+        temp_dir = pathlib.Path(temp_dir_name)
+
+        with contextlib.ExitStack() as stack:
+            try:
+                wheel = stack.enter_context(zipfile.ZipFile(wheel_file))
+            except FileNotFoundError:
+                raise ApplicationError(f"Missing wheel for version {version}: {wheel_file}") from None
+
+            wheel.extractall(temp_dir)
```
Does this just smoke-test that unpacking won't fail? Why not use the same mechanism pip would use, then?
```suggestion
        run('wheel', 'unpack', str(wheel_file), cwd=temp_dir)
```

mattclay replied:
There's already a separate check using `pip install` to make sure the wheel (and sdist) can be installed.

This check is done first using `extractall`, where the error handling is easier than using a subprocess. Having the contents in a temporary directory will make adding additional checks in the future easier (something already done for the sdist test).

webknjaz commented on packaging/release.py:
```diff
@@ -0,0 +1,1410 @@
+@contextlib.contextmanager
+def suppress_when(error_as_warning: bool) -> None:
+    """Conditionally convert an ApplicationError in the provided context to a warning."""
+    if error_as_warning:
+        try:
+            yield
+        except ApplicationError as ex:
+            display.warning(ex)
+    else:
+        yield
```
Let's dedent this a little bit:
```suggestion
    suppressed_exceptions = ApplicationError if error_as_warning else ()
    try:
        yield
    except suppressed_exceptions as ex:
        display.warning(ex)
```

mattclay replied:
While I'm not fond of the current approach of using `yield` twice, I think I find it more readable than using a variable to hold the exception type in this case.

webknjaz commented on packaging/release.py:
```diff
@@ -0,0 +1,1410 @@
+
+    commit = get_commit()
+
+    if commit != upstream_commit:
+        with suppress_when(allow_stale):
+            raise ApplicationError(f"The current commit ({commit}) does not match {upstream_ref} ({upstream_commit}).")
```
```suggestion
    with suppress_when(commit != upstream_commit and allow_stale):
        raise ApplicationError(f"The current commit ({commit}) does not match {upstream_ref} ({upstream_commit}).")
```

mattclay replied:
The suggested change breaks the intended error checking behavior.

webknjaz replied:
how?

mattclay replied:
It always generates a warning or an error, regardless of whether the commit matches or not.

webknjaz replied:
Oh... I've re-read the `suppress_when()` implementation and now I see what you mean. Thanks for the explanation.

webknjaz commented on packaging/release.py:
```diff
@@ -0,0 +1,1410 @@
+
+    dirty = git("status", "--porcelain", "--untracked-files=all", capture_output=True).stdout.strip().splitlines()
+
+    if dirty:
+        with suppress_when(allow_dirty):
+            raise ApplicationError(f"There are {len(dirty)} files which are untracked and/or have changes, which will be omitted from the build.")
```
```suggestion
    with suppress_when(dirty and allow_dirty):
        raise ApplicationError(f"There are {len(dirty)} files which are untracked and/or have changes, which will be omitted from the build.")
```

mattclay replied:
The suggested change breaks the intended error checking behavior.

webknjaz commented on packaging/release.py:
```diff
@@ -0,0 +1,1410 @@
+        url = f"mailto:{to}?{query_string}"
+
+        display.show("Opening email client through default web browser ...")
+        webbrowser.open(url)
```
This is something I'd avoid by default. I don't like things opening tabs in my browser unconditionally. Specifically, because I tend to have a lot of tabs and this may cause crashes when I don't expect, which usually leads to having to spend time restoring the browsing sessions. So I'd say this would be better as opt-in, not opt-out.

mattclay replied:
The goal was to keep the command invocations as simple as possible for the common use cases. Browser tabs are also opened for creating PRs and releases, so it makes sense to default to one for release announcements as well.

webknjaz replied:
I'd like to be able to disable all of the occurrences. Having this on, would make the script virtually unusable for me.

webknjaz commented on packaging/release.py:
```diff
@@ -0,0 +1,1410 @@
+
+
+# endregion
+# region Templates
```
Can we put them into files?

mattclay replied:
If we end up splitting this into multiple files, breaking out the templates makes sense. Until then, I'll keep them inline.

webknjaz commented on packaging/release.py:
```diff
@@ -0,0 +1,1410 @@
+def get_release_artifact_details(repository: str, version: Version, validate: bool) -> list[ReleaseArtifact]:
+    """Return information about the release artifacts hosted on PyPI."""
+    endpoint = PYPI_ENDPOINTS[repository]
+    url = f"{endpoint}/ansible-core/{version}/json"
```
Can we put `ansible-core` and `ansible_core` into constants and reuse them?

mattclay replied:
We could, but with only 4 uses between them, I don't think it's going to help with readability. They're not likely to ever change either.

webknjaz replied:
This is what I usually embed into CI for assertions to make sure parts of the workflow stay compatible. There could be more uses for such vars.

webknjaz commented on packaging/release.py:
```diff
@@ -0,0 +1,1410 @@
+        env = os.environ.copy()
+        env.pop("PYTHONPATH", None)  # avoid interference from ansible being injected into the environment
+        env.update(
+            PATH=f"{venv_bin_dir}:{env['PATH']}",
```
Nit: `os.pathsep` is a bit more generic:
```suggestion
            PATH=os.pathsep.join((venv_bin_dir, env['PATH'])),
```

mattclay replied:
I've used `os.pathsep` in the relevant places.

webknjaz commented on packaging/release.py:
```diff
@@ -0,0 +1,1410 @@
+PIP_OPTIONS = (
+    "--require-virtualenv",
+    "--disable-pip-version-check",
+)
```
It'd be better to set these through env vars since individual CLI options are not propagated into the internal pip invocations (like when it makes isolated venvs for PEP 517 builds).

mattclay replied:
These have been moved into env vars.

webknjaz commented on packaging/release.py:
```diff
@@ -0,0 +1,1410 @@
+    """A fatal application error which will be shown without a traceback."""
+
+
+class CalledProcessError(Exception):
```
why not
```suggestion
class CalledProcessError(subprocess.CalledProcessError):
```
?

mattclay replied:
Extending `subprocess.CalledProcessError` doesn't help here, since we'd have to override all its attributes and methods, and that still wouldn't provide accurate type hints.

webknjaz commented on packaging/release.py:
```diff
@@ -0,0 +1,1410 @@
+            if failed:
+                display.show(f"!!! {label}", color=Display.RED)
+            else:
+                display.show(f"<== {label}", color=Display.BLUE)
```
Why didn't you go for
```suggestion
            display.show(f"!!! {label}", color=Display.RED)
            raise

        display.show(f"<== {label}", color=Display.BLUE)
```
?

A flag changed in “except” and re-accessed in “finally” almost after the function's control flow is interrupted is probably not the easiest to follow.

Plus, this variant is less nested.

mattclay replied:
I reworked this a few times, and somehow missed this cleanup when doing so. It's simplified now.

webknjaz commented on packaging/release.py:
```diff
@@ -0,0 +1,1410 @@
+        webbrowser.open(url)
+    else:
+        print(f"TO: {recipients}")
+        print(f"SUBJECT: {message.subject}\n{message.body}")
```
How about having a visual empty line?
```suggestion
        print(f"SUBJECT: {message.subject}\n\n{message.body}")
```

mattclay replied:
There's a blank line before the body now.

webknjaz commented on packaging/release.py:
```diff
@@ -0,0 +1,1410 @@
+            PATH=f"{venv_bin_dir}:{env['PATH']}",
+        )
+
+        run("pip", "install", path, *PIP_OPTIONS, env=env, cwd=CHECKOUT_DIR)
```
I think I'd prefer using full paths for this and following commands, instead of relying on `$PATH`:
```suggestion
        run(str(venv_bin_dir / "pip"), "install", path, *PIP_OPTIONS, env=env, cwd=CHECKOUT_DIR)
```
(or make a helper wrapping `run()` appending the venv path)

mattclay replied:
I don't see any issue with relying on `PATH` here, when it's also necessary for other commands we invoke. I did experiment with some convenience methods for invoking commands in a venv, but decided it was more trouble than it was worth for such a simple use case.

webknjaz commented on packaging/release.py:
```diff
@@ -0,0 +1,1410 @@
+        if ex.status == http.HTTPStatus.NOT_FOUND:
+            raise ApplicationError(f"Version {version} not found on PyPI.") from None
+
+        raise Exception(f"Failed to get {version} from PyPI: {ex}") from ex
```
let's try using more accurate exceptions
```suggestion
        raise RuntimeError(f"Failed to get {version} from PyPI: {ex}") from ex
```

mattclay replied:
All uses of `raise Exception` have been replaced with `raise RuntimeError`.

webknjaz reviewed (commented):
Here's an incomplete first impression.

sivel commented on packaging/release.py:
```diff
@@ -0,0 +1,1410 @@
+    "ansible-announce@googlegroups.com",
+    "ansible-project@googlegroups.com",
+    "ansible-devel@googlegroups.com",
+)
```
Since we're using mailto, this is of course editable before sending, but just as an FYI, for pre-releases we've only been sending to ansible-devel.  Only GAs have been going to all 3.

mattclay replied:
I've split the recipient lists into two, one for final releases and one for development releases.

sivel commented on packaging/release.py:
```diff
@@ -0,0 +1,1410 @@
+Subject to the need for additional alpha releases, the first release beta is scheduled for X.
+{% else %}
+The next release candidate is planned to be released on {{ next_rc.strftime('%-d %B %Y') }}.
+The next general availability release will be one week after.
```
A nitpick on this line, since I wrote it, and had been too lazy to do anything about it, I always delete the newline and move it up.  Could add an extra newline too.  The unnecessary newline was probably just for line length when I did it.

mattclay replied:
I've removed the extra line break.

mattclay committed:
Add safety check to verify remotes are different

This will detect the scenario where the `devel` branch points at the remote the user's fork was forked from, instead of the user's fork.

mattclay committed:
Simplify command error handling

mattclay committed:
Separate final and pre-release recipient lists

mattclay committed:
Use os.pathsep to join PATH entries

mattclay committed:
Add blank line before message with `--no-mailto`

mattclay committed:
Remove unnecessary line break between sentences

mattclay committed:
Raise RuntimeError instead of Exception

mattclay committed:
Specify common pip options using env vars

mattclay committed:
Fix display of CalledProcessError

mattclay committed:
Use version specific porting guide

mattclay committed:
Fix path handling when .. is used

mattclay committed:
Match PyPI artifact terminology

webknjaz commented on packaging/release.py:
```diff
@@ -0,0 +1,1420 @@
+@functools.cache
+def get_remotes() -> Remotes:
+    """Return details about the remotes we need to use."""
+    # assume the devel branch has its upstream remote pointing to the user's fork
+    fork_remote_name = git("branch", "--list", "devel", "--format=%(upstream:remotename)", capture_output=True).stdout.strip()
```
This assumption is broken for me:
```console
$ git branch --list devel '--format=%(upstream:remotename)'
origin

$ git branch --list packaging/pep517-in-tree-backend-mutation-tests '--format=%(upstream:remotename)'
fork
```

mattclay replied:
@webknjaz Your `devel` branch remote points at ansible/ansible instead of your fork?

webknjaz replied:
Yep. This is because my userspace automation clones a lot of repos in a loop from their canonical locations, so they all have `origin` pointing upstream: https://github.com/webknjaz/ansible-gentoo-laptop/blob/e71ff8f/roles/gentoo-userspace/defaults/main.yml#L252-L573. It then adds a `fork` remote to select repos. It's worked like this for the last 5 years and is duplicated across different systems I have...
So I always push to `fork` while pulling from `origin` upstream, unless `origin` is where I work directly, that is.

webknjaz replied:
Maybe, we could implement having a setting in `.git/config`? I've implemented some helpers for this @ https://github.com/python/cherry-picker/blob/2c1f6b9/cherry_picker/cherry_picker.py#L995-L1022.

mattclay replied:
@webknjaz There's no issue with the `origin` remote pointing upstream to the `ansible/ansible` repo. Based on your description, it sounds like you never push to your `devel` branch, since that wouldn't go to your fork. Is that correct?

webknjaz replied:
Sometimes I sync the fork via something like `git fetch --all && git push fork origin/devel:devel`.
